src/main/java/dev/openfeature/sdk/ImmutableContext.java

```diff
@@ -0,0 +1,83 @@
+package dev.openfeature.sdk;
+
+import lombok.Getter;
+import lombok.ToString;
+import lombok.experimental.Delegate;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * The EvaluationContext is a container for arbitrary contextual data
+ * that can be used as a basis for dynamic evaluation.
+ * The ImmutableContext is an EvaluationContext implementation which is threadsafe, and whose attributes can
+ * not be modified after instantiation.
+ */
+@ToString
+@SuppressWarnings("PMD.BeanMembersShouldSerialize")
+public final class ImmutableContext implements EvaluationContext {
+
+    @Getter
+    private final String targetingKey;
+    @Delegate
+    private final Structure structure;
+
+    /**
+     * create an immutable context with an empty targeting_key and attributes provided.
+     */
+    public ImmutableContext() {
+        this("", new HashMap<>());
+    }
+
+    /**
+     * create an immutable context with an attributes provided.
+     *
+     * @param attributes evaluation context attributes
+     */
+    public ImmutableContext(Map<String, Value> attributes) {
+        this("", attributes);
+    }
+
+    /**
+     * create an immutable context with given targetingKey and attributes provided.
+     *
+     * @param targetingKey targeting key
+     * @param attributes   evaluation context attributes
+     */
+    public ImmutableContext(String targetingKey, Map<String, Value> attributes) {
+        this.structure = new ImmutableStructure(attributes);
+        this.targetingKey = targetingKey;
+    }
+
+    @Override
+    public void setTargetingKey(String targetingKey) {
+        throw new UnsupportedOperationException("changing of targeting key is not allowed");
+    }
```

Comment:
@justinabrahms @Kavindu-Dodan 

This is unfortunate. I don' think the `EvaluationContext` interface should have `setTargetingKey`, since that makes it mutable. I would like to remove it from that interface, but that would be a breaking change (though minor). How do you all feel about that?

Technically the "evaluation context" aspects of the spec are still experimental, so we could implement this breaking change with only a minor version update, as per our documentation, but I would require changes from authors in cases they declare `EvaluationContext`s and then call this method (they would have to update to use the subclass).

@justinabrahms @beeme1mr @Kavindu-Dodan @thiyagu06

Reply:
I think I agree with this.

Reply:
I will open a new issue for this, then resolve this conversation.

Reply:
I've created [this issue](https://github.com/open-feature/java-sdk/issues/253#issue-1538452873).